File: apps/web/lib/media-access.ts
import "server-only";

import { readFile } from "node:fs/promises";
import { relative, resolve, sep } from "node:path";

import { withAuditedProjectWrite } from "@/lib/audit";
import { requireCurrentProjectPermission } from "@/lib/data";
import { query } from "@/lib/db";
import {
  assertMediaActionAllowed,
  MediaAccessSigner,
  type MediaAccessAction,
  safeDownloadName
} from "@/lib/media-access-core";
import { correlationId } from "@/lib/observability";
import { getWebRuntimeConfig } from "@/lib/runtime-config";

type AssetAccessRow = {
  id: number;
  projectId: number;
  storageKey: string;
  mimeType: string | null;
  kind: string;
  fileName: string | null;
  sensitive: boolean;
};

async function loadAsset(projectId: number, assetId: number) {
  const result = await query<AssetAccessRow>(
    `select asset.id, asset.project_id as "projectId", asset.storage_key as "storageKey",
            asset.mime_type as "mimeType", asset.kind,
            intent.file_name as "fileName",
            (coalesce(asset.metadata_json->>'sensitive', 'false') = 'true' or exists (
              select 1 from evidence_links evidence
               where evidence.project_id = asset.project_id and evidence.asset_id = asset.id
                 and evidence.is_published
            )) as sensitive
       from assets asset
       left join asset_upload_intents intent
         on intent.project_id = asset.project_id and intent.asset_id = asset.id
      where asset.project_id = $1 and asset.id = $2 and asset.status = 'available'`,
    [projectId, assetId]
  );
  if (!result.rows[0]) throw new Error("ASSET_NOT_FOUND");
  return result.rows[0];
}

export async function issueMediaAccess(
  projectId: number,
  assetId: number,
  action: MediaAccessAction,
  requestId?: string | null
) {
  const { user, access } = await requireCurrentProjectPermission(projectId, "project:view");
  const asset = await loadAsset(projectId, assetId);
  assertMediaActionAllowed({ action, role: access.role, permissions: access.permissions, sensitive: asset.sensitive });
  const issue = () => new MediaAccessSigner(getWebRuntimeConfig().authSecret).issue({
    projectId, assetId, action, ttlSeconds: 120
  });
  if (action === "download" && asset.sensitive) {
    return withAuditedProjectWrite(
      {
        projectId, teamId: access.teamId, requestId: correlationId(requestId), actorUserId: user.id,
        action: "media.sensitive_download", resourceType: "asset", resourceId: String(assetId),
        input: { action, assetId }, policyResult: { permission: "project:view", sensitive: true }
      },
      async () => issue()
    );
  }
  return issue();
}

export async function readAuthorizedMediaContent(input: {
  projectId: number;
  assetId: number;
  action: MediaAccessAction;
  expires: string;
  signature: string;
}) {
  const { access } = await requireCurrentProjectPermission(input.projectId, "project:view");
  const asset = await loadAsset(input.projectId, input.assetId);
  assertMediaActionAllowed({
    action: input.action, role: access.role, permissions: access.permissions, sensitive: asset.sensitive
  });
  const verified = new MediaAccessSigner(getWebRuntimeConfig().authSecret).verify(input);
  if (!verified) throw new Error("MEDIA_ACCESS_TOKEN_INVALID");

  const root = getWebRuntimeConfig().objectStorageLocalRoot;
  if (!root) throw new Error("OBJECT_STORAGE_UNAVAILABLE");
  if (!asset.storageKey.startsWith(`projects/${input.projectId}/`) || asset.storageKey.includes("\\")) {
    throw new Error("INVALID_ASSET_STORAGE_KEY");
  }
  const absoluteRoot = resolve(root);
  const objectPath = resolve(absoluteRoot, asset.storageKey);
  const scopedPath = relative(absoluteRoot, objectPath);
  if (scopedPath === ".." || scopedPath.startsWith(`..${sep}`)) throw new Error("INVALID_ASSET_STORAGE_KEY");
  const body = await readFile(objectPath);
  return {
    body,
    contentType: asset.mimeType ?? "application/octet-stream",
    disposition: input.action === "download"
      ? `attachment; filename="${safeDownloadName(asset.fileName, `asset-${asset.id}`)}"`
      : "inline"
  };
}

File: apps/server/internal/httpapi/media_access.go
package httpapi

import (
	"aerosight/server/internal/database"
	"aerosight/server/internal/database/sqlcgen"
	"aerosight/server/internal/httptransport"
	"aerosight/server/internal/media"
	"context"
	"database/sql"
	"errors"
	"fmt"
	"strconv"
	"time"

	"github.com/gin-gonic/gin"
)

func (s *Server) AttachMediaStorage(root string) { s.mediaStorageRoot = root }
func (s *Server) mediaAccessRoutes() {
	s.router.POST("/api/projects/:id/assets/import", s.requireUser, s.importImageAsset)
	group := s.router.Group("/api/projects/:id/assets/:assetId", s.requireUser)
	group.GET("/access", s.timeout, s.issueMediaAccess)
	group.GET("/content", s.readMediaContent)
	group.HEAD("/content", s.readMediaContent)
}
func mediaAccessScope(c *gin.Context) (int32, int32, string, error) {
	pid, err := projectID(c)
	if err != nil {
		return 0, 0, "", err
	}
	id, err := strconv.ParseInt(c.Param("assetId"), 10, 32)
	action := c.Query("action")
	if err != nil || id <= 0 || !media.ValidAccessAction(action) {
		return 0, 0, "", errors.New("INVALID_MEDIA_ACCESS_INPUT")
	}
	return pid, int32(id), action, nil
}
func (s *Server) mediaAsset(ctx context.Context, q *sqlcgen.Queries, uid, pid, aid int32, action string) (sqlcgen.ReadMediaAccessAssetRow, sqlcgen.GetProjectAccessRow, error) {
	access, err := s.projectAccess(ctx, q, uid, pid, "project:view")
	if err != nil {
		return sqlcgen.ReadMediaAccessAssetRow{}, access, err
	}
	asset, err := q.ReadMediaAccessAsset(ctx, sqlcgen.ReadMediaAccessAssetParams{ProjectID: pid, ID: aid})
	if err != nil {
		return asset, access, err
	}
	if !media.AccessAllowed(action, access.Role, effectivePermissions(access.Role, access.Permissions), asset.Sensitive) {
		return asset, access, errors.New("SENSITIVE_MEDIA_DOWNLOAD_DENIED")
	}
	return asset, access, nil
}
func (s *Server) readMediaAsset(ctx context.Context, uid, pid, aid int32, action string) (sqlcgen.ReadMediaAccessAssetRow, sqlcgen.GetProjectAccessRow, error) {
	tx, err := s.db.BeginTx(ctx, &sql.TxOptions{Isolation: sql.LevelRepeatableRead, ReadOnly: true})
	if err != nil {
		return sqlcgen.ReadMediaAccessAssetRow{}, sqlcgen.GetProjectAccessRow{}, err
	}
	defer tx.Rollback()
	asset, access, err := s.mediaAsset(ctx, s.queries.WithTx(tx), uid, pid, aid, action)
	if err != nil {
		return asset, access, err
	}
	err = tx.Commit()
	return asset, access, err
}
func (s *Server) issueMediaAccess(c *gin.Context) {
	fail := func() { s.failure(c, 403, "Unable to access media") }
	pid, aid, action, err := mediaAccessScope(c)
	if err != nil {
		fail()
		return
	}
	ctx := c.Request.Context()
	uid := currentUser(c).ID
	asset, access, err := s.readMediaAsset(ctx, uid, pid, aid, action)
	if err != nil {
		fail()
		return
	}
	issue := func() (media.Access, error) {
		return media.IssueAccess(s.credentialSecret, pid, aid, action, time.Now(), 120)
	}
	var result media.Access
	if action == "download" && asset.Sensitive {
		audit := database.AuditContext{ProjectID: pid, TeamID: access.TeamID, ActorUserID: uid, RequestID: c.GetHeader("X-Request-ID"), Action: "media.sensitive_download", ResourceType: "asset", ResourceID: strconv.FormatInt(int64(aid), 10), Input: gin.H{"action": action, "assetId": aid}, PolicyResult: map[string]any{"permission": "project:view", "sensitive": true}}
		result, err = database.AuditedWrite(ctx, s.db, audit, s.authorizeWrite(uid, pid, access.TeamID, "project:view", false), func(w *database.WriteTx) (media.Access, error) {
			if _, _, e := s.mediaAsset(ctx, w.Queries, uid, pid, aid, action); e != nil {
				return media.Access{}, e
			}
			return issue()
		})
	} else {
		result, err = issue()
	}
	if err != nil {
		fail()
		return
	}
	c.JSON(200, result)
}
func (s *Server) readMediaContent(c *gin.Context) {
	fail := func() { s.failure(c, 403, "Media unavailable") }
	pid, aid, action, err := mediaAccessScope(c)
	if err != nil {
		fail()
		return
	}
	lookup, cancel := context.WithTimeout(c.Request.Context(), s.cfg.RequestTimeout)
	asset, _, err := s.readMediaAsset(lookup, currentUser(c).ID, pid, aid, action)
	lookupError := lookup.Err()
	cancel()
	if lookupError == context.DeadlineExceeded {
		s.failure(c, 504, "REQUEST_TIMEOUT")
		return
	}
	if err != nil {
		fail()
		return
	}
	if !media.VerifyAccess(s.credentialSecret, pid, aid, action, c.Query("expires"), c.Query("signature"), time.Now()) {
		fail()
		return
	}
	file, err := media.OpenProjectObject(s.mediaStorageRoot, pid, asset.StorageKey)
	if err != nil {
		fail()
		return
	}
	defer file.Close()
	contentType := "application/octet-stream"
	if asset.MimeType.Valid {
		contentType = asset.MimeType.String
	}
	disposition := "inline"
	if action == "download" {
		var name *string
		if asset.FileName.Valid {
			name = &asset.FileName.String
		}
		disposition = fmt.Sprintf(`attachment; filename="%s"`, media.SafeDownloadName(name, fmt.Sprintf("asset-%d", aid)))
	}
	c.Header("Content-Type", contentType)
	c.Header("Content-Disposition", disposition)
	c.Header("Cache-Control", "private, no-store")
	c.Header("X-Content-Type-Options", "nosniff")
	// No Last-Modified validators: each access is reauthorized and must not turn
	// into a shared/public cache hit. ServeContent implements byte ranges/HEAD.
	httptransport.ServeContent(c.Writer, c.Request, file)
}
